test/integration/daemonset/daemonset_test.go

```diff
@@ -1159,3 +1159,101 @@ func TestUpdateStatusDespitePodCreationFailure(t *testing.T) {
 		validateDaemonSetStatus(dsClient, ds.Name, int32(limitedPodNumber), t)
 	})
 }
+
+func TestDaemonSetRollingUpdateWithTolerations(t *testing.T) {
+	ctx, closeFn, dc, informers, clientset := setup(t)
+	defer closeFn()
+	ns := framework.CreateNamespaceOrDie(clientset, "daemonset-rollingupdate-with-tolerations-test", t)
+	defer framework.DeleteNamespaceOrDie(clientset, ns, t)
+
+	dsClient := clientset.AppsV1().DaemonSets(ns.Name)
+	podClient := clientset.CoreV1().Pods(ns.Name)
+	nodeClient := clientset.CoreV1().Nodes()
+	podInformer := informers.Core().V1().Pods().Informer()
+
+	informers.Start(ctx.Done())
+	go dc.Run(ctx, 2)
+
+	zero := intstr.FromInt32(0)
+	maxSurge := intstr.FromInt32(1)
+	ds := newDaemonSet("foo", ns.Name)
+	ds.Spec.UpdateStrategy = apps.DaemonSetUpdateStrategy{
+		Type: apps.RollingUpdateDaemonSetStrategyType,
+		RollingUpdate: &apps.RollingUpdateDaemonSet{
+			MaxUnavailable: &zero,
+			MaxSurge:       &maxSurge,
+		},
+	}
+	defer cleanupDaemonSets(t, clientset, ds)
+
+	// Add four nodes with zone-y, zone-z or common taint
+	var taints []v1.Taint
+	var node *v1.Node
+	for i := 0; i < 6; i++ {
+		if i < 2 {
+			taints = []v1.Taint{
+				{Key: "zone-y", Effect: v1.TaintEffectNoSchedule},
+			}
+		} else if i < 4 {
+			taints = []v1.Taint{
+				{Key: "zone-z", Effect: v1.TaintEffectNoSchedule},
+			}
+		} else {
+			taints = []v1.Taint{
+				{Key: "common", Effect: v1.TaintEffectNoSchedule},
+			}
+		}
+		node = newNode(fmt.Sprintf("node-%d", i), nil)
+		node.Spec.Taints = taints
+		_, err := nodeClient.Create(context.TODO(), node, metav1.CreateOptions{})
+		if err != nil {
+			t.Fatalf("Failed to create node: %v", err)
+		}
+	}
+
+	// Create DaemonSet with zone-y toleration
+	ds.Spec.Template.Spec.Tolerations = []v1.Toleration{
+		{Key: "zone-y", Operator: v1.TolerationOpExists},
+		{Key: "common", Operator: v1.TolerationOpExists},
+	}
+	_, err := dsClient.Create(ctx, ds, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to create DaemonSet: %v", err)
+	}
+	validateDaemonSetPodsAndMarkReady(podClient, podInformer, 4, t)
+	validateDaemonSetStatus(dsClient, ds.Name, 4, t)
+
+	// Update DaemonSet with zone-z toleration
+	ds.Spec.Template.Spec.Tolerations = []v1.Toleration{
+		{Key: "zone-z", Operator: v1.TolerationOpExists},
+		{Key: "common", Operator: v1.TolerationOpExists},
+	}
+	_, err = dsClient.Update(ctx, ds, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to update DaemonSet: %v", err)
+	}
+	validateDaemonSetPodsAndMarkReady(podClient, podInformer, 4, t)
```

Comment:
we should first check that we have 6 pods with `validateDaemonSetPodsActive` and then mark our pods ready

Reply:
@atiratree 
OK, I'll add that check.

Isn't expected numberPods of `validateDaemonSetPodsActive` 4 when first create DaemonSet with zone-y and common toleration?
(2 pods are on Zone Y nodes, 2 pods are on Common nodes)

And I found that expected numberPods of `validateDaemonSetPodsActive` is 7 when update DaemonSet with zone-z and common toleration.

It's because: 
- daemon_controller `manage()` creates a pod on each node where it should run and dosen't exist, regardless of `maxSurge`.
  - New 2 pods are created and Pending on Zone Z nodes
- `remainingSurge` will be calculated based on schedulable nodes in `rollingUpdate`
  - New 1 pod are created as surge  and Pending on Zone Common node
  - Old 2 pods on Zone Common nodes will remain existing and Running 
- Pods that violate scheduling constraints will be removed in `rollingUpdate` when `numAvailable > desiredNumberScheduled` is met. 
  - Old 2 pods on Zone Y nodes will remain existing and Running until other new pods become available after `validateDaemonSetPodsAndMarkReady`

What do you think about that?

Reply:
yup 4 when we create and a good explanation that we will have 7pods when we update the daemon set. Maybe you can write a short comment explaining this in the code?